pkg/storageincentives/redistributionstate.go

```diff
@@ -0,0 +1,151 @@
+// Copyright 2023 The Swarm Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package storageincentives
+
+import (
+	"context"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethersphere/bee/pkg/log"
+	"github.com/ethersphere/bee/pkg/settlement/swap/erc20"
+	"github.com/ethersphere/bee/pkg/storage"
+	"github.com/ethersphere/bee/pkg/storageincentives/redistribution"
+	"github.com/ethersphere/bee/pkg/swarm"
+	"math/big"
+	"sync"
+)
+
+const redistributionStatusKey = "redistribution_state"
+const loggerNameNode = "nodestatus"
+
+type RedistributionState struct {
+	stateStore     storage.StateStorer
+	erc20Service   erc20.Service
+	logger         log.Logger
+	overlay        swarm.Address
+	mtx            sync.Mutex
+	status         Status
+	currentBalance *big.Int
+	contract       redistribution.Contract
+}
+
+// Status provide internal status of the nodes in the redistribution game
+type Status struct {
+	Phase           PhaseType
+	IsFrozen        bool
+	Round           uint64
+	LastWonRound    uint64
+	LastPlayedRound uint64
+	LastFrozenRound uint64
+	Block           uint64
+	Reward          *big.Int
+	Fees            *big.Int
+}
+
+func NewRedistributionState(logger log.Logger, stateStore storage.StateStorer, erc20Service erc20.Service, contract redistribution.Contract) RedistributionState {
+	return RedistributionState{
+		stateStore:     stateStore,
+		erc20Service:   erc20Service,
+		logger:         logger.WithName(loggerNameNode).Register(),
+		currentBalance: big.NewInt(0),
+		contract:       contract,
+		status: Status{
+			Round:  0,
+			Block:  0,
+			Reward: nil,
+			Fees:   big.NewInt(0),
+		},
+	}
+}
+
+func (n *RedistributionState) SetCurrentEvent(p PhaseType, r uint64, b uint64) {
+	n.mtx.Lock()
+	defer n.mtx.Unlock()
+	n.status.Phase = p
+	n.status.Round = r
+	n.status.Block = b
+	n.saveStatus()
+}
+
+func (n *RedistributionState) IsFrozen(f bool) {
+	n.mtx.Lock()
+	defer n.mtx.Unlock()
+	n.status.IsFrozen = f
+	n.saveStatus()
+}
```

Comment:
would it make sense for `IsFrozen` to accept additional parameter `round` so this method could also set `LastFrozenRound` if frozen? if so then `SetLastFrozenRound` method would not be needed.

would name `SetFrozen` be more convenient for this method (following same convention)?

Reply:
Good suggestion.